### ISSUE TITLE
Gradient of output wrt input

### ISSUE THREAD

**Issue by FlashTek** [Microsoft/CNTK#2416]
I would like to calculate the gradient of my network's output with respect to one of the inputs and evaluate this gradient for a specific set of input variables (using C++). Therefore, I thought about using the `outputFunctionPtr->Gradients()` function like this:

    auto gradientArguments = {{input1FunctionPtr, input1ValuePtr}, {input2FunctionPtr, input2ValuePtr}}
    auto gradientOutput = {{outputFunctionPtr, nullptr}}
    outputFunctionPtr->Gradients(gradientArguments, gradientOutput, device) 

sadly, this fails with this error message:

> Gradient requested for Variable 'Output('output', [1], [*, #])' which is not a leaf input (Input, Parameter or Constant) of the Function 'Composite(Plus): Input('input1', [2], [*, #]), Input('inpu2t', [1], [*, #]) -> Output('output', [1], [*, #])'; this is currently unsupported.'

Therefore, I wondered, if I am just using a wrong function to calculate the gradient, of if it even is not possible to do something like this at the moment.

Edit: I am looking for something which behaves similiar to tensorflows `tf.gradients` method.

**Comment by n17s** [comment]
Gradients can be computed for Parameters or for InputVariables if they are constructed with `needsGradient=true`. Intermediate results such as in your example are not supported and because of memory sharing they may even contain garbage values even after a simple forward propagation.

To compute the gradient of a network that was constructed as `y=f(x); z=g(y)` with respect to `z` you will need to:
- make two networks; one for `f` and one for `g`, and make the input to `g` an `InputVariable` with `needsGradient=true`.
- Evaluate the network for `f`, take the output `ValuePtr`, and provide it as the value for the `g`'s input. Then call Gradients() on `g`.

**Comment by FlashTek** [comment]
What do you mean with `values even after a simple forward propagation`? Why should the forward propagation fail on a network with two inputs and one output?

>To compute the gradient of a network that was constructed as y=f(x); z=g(y) with respect to z

And it is not possible to calculate `dz/dx` in an way at all?

**Comment by xqustc** [comment]
@n17s Hi, how to call Gradients() ? I just found C.grad() function. How to set the parameter "wrt" in grad() if I want to get the gradients with respect to the output neural network？
